src/optimization/steepest_descent_method.cpp

```diff
@@ -0,0 +1,302 @@
+#include <k52/parallel/mpi/constants.h>
+
+#endif
+
+#include <iostream>
+#include <cmath>
+
+#include <k52/optimization/params/i_continuous_parameters.h>
+
+using ::std::vector;
+
+namespace k52
+{
+namespace optimization
+{
+
+SteepestDescentMethod::SteepestDescentMethod(
+	double alpha,
+	double increment_of_the_argument,
+	size_t max_iteration_number,
+	double precision)
+	: alpha_(alpha)
+	, increment_of_the_argument_(increment_of_the_argument)
+	, max_iteration_number_(max_iteration_number)
+	, precision_(precision)
+{
+}
+
+SteepestDescentMethod* SteepestDescentMethod::Clone() const
+{
+	return new SteepestDescentMethod(
+		alpha_,
+		increment_of_the_argument_,
+		max_iteration_number_,
+		precision_);
+}
+
+std::string SteepestDescentMethod::get_name() const
+{
+	return "Steepest Descent Method";
+}
+
+
+#ifdef BUILD_WITH_MPI
+void SteepestDescentMethod::Send(boost::mpi::communicator* communicator, int target) const
+ {
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+
+void SteepestDescentMethod::Receive(boost::mpi::communicator* communicator, int source)
+{
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+#endif
+
+
+vector<double> SteepestDescentMethod::FindOptimalParameters(const vector<double>& initial_parameters)
+{
+	vector<double> parameters(initial_parameters);
+	size_t counter = 0;
+	double lambda = 1; 
+
+	do
+	{
+		counter++;
+		vector<double> gradient = CalculateGradient(parameters);
+		vector<double> invert_gradiend = InvertVector(gradient);
+
+		lambda = GoldenSectionSearch(parameters, lambda, invert_gradiend);
+
+		vector<double> next_step_parameters(parameters.size());
+		for (size_t i = 0; i < next_step_parameters.size(); ++i)
+		{
+			next_step_parameters[i] = parameters[i] + lambda * invert_gradiend[i];
+		}
+
+		if (IsExitCriteriaFulfilled(next_step_parameters))
+		{
+			break;
+		}
+		
+		parameters = next_step_parameters;
+
+		
+		if (counter == max_iteration_number_)
+		{
+			std::cout << " Minimun has not found"<<std::endl;
+			break;
+		}
+	} while (true);
+
+	return parameters;
+}
+
+bool SteepestDescentMethod::IsExitCriteriaFulfilled(
+	vector<double>& steps_array) const
+{
+	double temp = 0;
+	vector<double> gradient = CalculateGradient(steps_array);
+	for (size_t i = 0; i < steps_array.size(); ++i)
+	{
+		temp += gradient[i] * gradient[i];
+	}
+	if (sqrt(temp) > precision_)
+	{
+		return false;
+	}
+	return true;
+}
+
+vector<double> SteepestDescentMethod::CalculateGradient(
+	const vector<double>& parameters) const
+{
```

Comment:
This will require a call  method "CountObjectiveFunctionValueToMinimize" from ContinuousOptimizer class, that is super class of SteepestDescentMethod. How to implement it? If passed as parameter - it's a little useful.